File: ZezeJava/src/Zeze/Net/Acceptor.java
package Zeze.Net;

import Zeze.*;

public class Acceptor {
	private Service Service;
	public final Service getService() {
		return Service;
	}
	private void setService(Service value) {
		Service = value;
	}
	private int Port = 0;
	public final int getPort() {
		return Port;
	}
	private String Ip = "";
	public final String getIp() {
		return Ip;
	}
	private AsyncSocket Socket;
	public final AsyncSocket getSocket() {
		return Socket;
	}
	private void setSocket(AsyncSocket value) {
		Socket = value;
	}
	public final String getName() {
		return String.format("%1$s:%2$s", getIp(), getPort());
	}

	public Acceptor(int port, String ip) {
		Port = port;
		Ip = ip;
	}

	public Acceptor(XmlElement self) {
		String attr = self.GetAttribute("Port");
		if (attr.length() > 0) {
			Port = Integer.parseInt(attr);
		}
		Ip = self.GetAttribute("Ip");
	}

	public final void SetService(Service service) {
		synchronized (this) {
			if (getService() != null) {
				throw new RuntimeException(String.format("Acceptor of '%1$s' Service != null", getName()));
			}
			setService(service);
		}
	}

	public final void Start() {
		synchronized (this) {
			if (null != getSocket()) {
				return;
			}

			setSocket(getIp().length() > 0 ? getService().NewServerSocket(getIp(), getPort()) : getService().NewServerSocket(System.Net.IPAddress.Any, getPort()));
			getSocket().setAcceptor(this);
		}
	}

	public final void Stop() {
		synchronized (this) {
			if (getSocket() != null) {
				getSocket().close();
			}
			setSocket(null);
		}
	}
}

File: Zege/Zeze/Net/Acceptor.cs
using System;
using System.Net;
using System.Xml;

namespace Zeze.Net
{
    public class Acceptor
    {
        public Service Service { get; private set; }
        public int Port;
        public string Ip;
        public AsyncSocket Socket { get; private set; }
        public string Name => $"{Ip}:{Port}";

        public Acceptor(int port, string ip)
        {
            Port = port;
            Ip = ip;
        }

        public Acceptor(XmlElement self)
        {
            string attr = self.GetAttribute("Port");
            if (attr.Length > 0)
                Port = int.Parse(attr);
            Ip = self.GetAttribute("Ip");
        }

        internal void SetService(Service service)
        {
            lock (this)
            {
                if (Service != null)
                    throw new Exception($"Acceptor of '{Name}' Service != null");
                Service = service;
            }
        }

        public void Start()
        {
            lock (this)
            {
                if (Socket != null)
                    return;

                Socket = Ip.Length > 0
                    ? Service.NewServerSocket(Ip, Port, this)
                    : Service.NewServerSocket(IPAddress.Any, Port, this);
            }
        }

        public void Stop()
        {
            lock (this)
            {
                Socket?.Dispose();
                Socket = null;
            }
        }
    }
}
